File: promptdev-backend/src/main/java/com/promptdev/service/BitbucketService.java
package com.promptdev.service;

import com.promptdev.config.BitbucketConfig;
import com.promptdev.dto.bitbucket.*;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.core.ParameterizedTypeReference;
import org.springframework.http.MediaType;
import org.springframework.stereotype.Service;
import org.springframework.web.client.RestClient;

import java.util.Collections;
import java.util.List;

@Service
public class BitbucketService {

    private static final Logger log = LoggerFactory.getLogger(BitbucketService.class);

    private final RestClient bitbucketRestClient;
    private final RestClient bitbucketBranchUtilsClient;
    private final BitbucketConfig bitbucketConfig;

    public BitbucketService(
            RestClient bitbucketRestClient,
            RestClient bitbucketBranchUtilsClient,
            BitbucketConfig bitbucketConfig) {
        this.bitbucketRestClient = bitbucketRestClient;
        this.bitbucketBranchUtilsClient = bitbucketBranchUtilsClient;
        this.bitbucketConfig = bitbucketConfig;
    }

    /**
     * Get repository details
     */
    public RepositoryResponse getRepository(String projectKey, String repoSlug) {
        log.info("Fetching repository: {}/{}", projectKey, repoSlug);
        return bitbucketRestClient.get()
                .uri("/projects/{projectKey}/repos/{repoSlug}", projectKey, repoSlug)
                .retrieve()
                .body(RepositoryResponse.class);
    }

    /**
     * List repositories in a project
     */
    public List<RepositoryResponse> listRepositories(String projectKey) {
        log.info("Listing repositories in project: {}", projectKey);
        PagedResponse<RepositoryResponse> response = bitbucketRestClient.get()
                .uri("/projects/{projectKey}/repos?limit=100", projectKey)
                .retrieve()
                .body(new ParameterizedTypeReference<>() {});
        return response != null ? response.values() : Collections.emptyList();
    }

    /**
     * Get the default branch of a repository
     */
    public BranchResponse getDefaultBranch(String projectKey, String repoSlug) {
        log.info("Fetching default branch for {}/{}", projectKey, repoSlug);
        return bitbucketRestClient.get()
                .uri("/projects/{projectKey}/repos/{repoSlug}/default-branch", projectKey, repoSlug)
                .retrieve()
                .body(BranchResponse.class);
    }

    /**
     * Create a new branch in the repository
     */
    public BranchResponse createBranch(String projectKey, String repoSlug, 
                                       String branchName, String startPoint) {
        log.info("Creating branch '{}' from '{}' in {}/{}", branchName, startPoint, projectKey, repoSlug);
        
        CreateBranchRequest request = new CreateBranchRequest(branchName, startPoint);
        
        return bitbucketRestClient.post()
                .uri("/projects/{projectKey}/repos/{repoSlug}/branches", projectKey, repoSlug)
                .contentType(MediaType.APPLICATION_JSON)
                .body(request)
                .retrieve()
                .body(BranchResponse.class);
    }

    /**
     * Create a pull request
     */
    public PullRequestResponse createPullRequest(String projectKey, String repoSlug,
                                                  String title, String description,
                                                  String sourceBranch, String targetBranch,
                                                  List<String> reviewers) {
        log.info("Creating PR from '{}' to '{}' in {}/{}", sourceBranch, targetBranch, projectKey, repoSlug);

        List<CreatePullRequestRequest.ReviewerInfo> reviewerInfos = reviewers != null 
                ? reviewers.stream()
                    .map(r -> new CreatePullRequestRequest.ReviewerInfo(
                            new CreatePullRequestRequest.UserInfo(r)))
                    .toList()
                : Collections.emptyList();

        CreatePullRequestRequest request = new CreatePullRequestRequest(
                title,
                description,
                new CreatePullRequestRequest.RefInfo("refs/heads/" + sourceBranch),
                new CreatePullRequestRequest.RefInfo("refs/heads/" + targetBranch),
                reviewerInfos
        );

        return bitbucketRestClient.post()
                .uri("/projects/{projectKey}/repos/{repoSlug}/pull-requests", projectKey, repoSlug)
                .contentType(MediaType.APPLICATION_JSON)
                .body(request)
                .retrieve()
                .body(PullRequestResponse.class);
    }

    /**
     * Get a pull request by ID
     */
    public PullRequestResponse getPullRequest(String projectKey, String repoSlug, Long prId) {
        log.info("Fetching PR #{} from {}/{}", prId, projectKey, repoSlug);
        return bitbucketRestClient.get()
                .uri("/projects/{projectKey}/repos/{repoSlug}/pull-requests/{prId}", 
                        projectKey, repoSlug, prId)
                .retrieve()
                .body(PullRequestResponse.class);
    }

    /**
     * List branches in a repository
     */
    public List<BranchResponse> listBranches(String projectKey, String repoSlug, String filterText) {
        log.info("Listing branches in {}/{}", projectKey, repoSlug);
        boolean hasFilter = filterText != null && !filterText.isBlank();
        String uri = hasFilter
                ? "/projects/{projectKey}/repos/{repoSlug}/branches?filterText={filter}&limit=100"
                : "/projects/{projectKey}/repos/{repoSlug}/branches?limit=100";

        PagedResponse<BranchResponse> response = hasFilter
                ? bitbucketRestClient.get()
                        .uri(uri, projectKey, repoSlug, filterText)
                        .retrieve()
                        .body(new ParameterizedTypeReference<>() {})
                : bitbucketRestClient.get()
                        .uri(uri, projectKey, repoSlug)
                        .retrieve()
                        .body(new ParameterizedTypeReference<>() {});
        return response != null ? response.values() : Collections.emptyList();
    }

    /**
     * Delete a branch (for cleanup if needed)
     */
    public void deleteBranch(String projectKey, String repoSlug, String branchName) {
        log.info("Deleting branch '{}' in {}/{}", branchName, projectKey, repoSlug);
        
        // Branch delete uses branch-utils API
        bitbucketBranchUtilsClient.delete()
                .uri(uriBuilder -> uriBuilder
                        .path("/projects/{projectKey}/repos/{repoSlug}/branches")
                        .build(projectKey, repoSlug))
                .header("Content-Type", "application/json")
                .retrieve()
                .toBodilessEntity();
    }

    /**
     * Get clone URL for a repository
     */
    public String getCloneUrl(String projectKey, String repoSlug) {
        RepositoryResponse repo = getRepository(projectKey, repoSlug);
        String httpUrl = repo.getCloneUrl("http");
        if (httpUrl != null) {
            return httpUrl;
        }
        // Fallback: construct URL
        return bitbucketConfig.getBaseUrl() + "/scm/" + projectKey.toLowerCase() + "/" + repoSlug + ".git";
    }

    /**
     * Construct the web URL for a pull request
     */
    public String getPullRequestWebUrl(String projectKey, String repoSlug, Long prId) {
        return String.format("%s/projects/%s/repos/%s/pull-requests/%d",
                bitbucketConfig.getBaseUrl(), projectKey, repoSlug, prId);
    }
}

File: promptdev-ui/lib/services/bitbucket-service.ts
/**
 * Bitbucket Server API service.
 * Direct HTTP calls to Bitbucket Server REST API.
 */

export function getBitbucketConfig() {
  const baseUrl = process.env.BITBUCKET_URL;
  const username = process.env.BITBUCKET_USERNAME;
  const token = process.env.BITBUCKET_TOKEN;
  return { baseUrl, username, token };
}

function getAuthHeaders(): Record<string, string> {
  const { token } = getBitbucketConfig();
  if (!token) return {};
  return { Authorization: `Bearer ${token}` };
}

async function bitbucketFetch<T>(path: string, options?: RequestInit): Promise<T> {
  const { baseUrl } = getBitbucketConfig();
  if (!baseUrl) throw new Error("BITBUCKET_URL is not configured");

  const url = `${baseUrl}/rest/api/latest${path}`;
  const response = await fetch(url, {
    ...options,
    headers: {
      "Content-Type": "application/json",
      ...getAuthHeaders(),
      ...options?.headers,
    },
  });

  if (!response.ok) {
    const text = await response.text();
    throw new Error(`Bitbucket API error ${response.status}: ${text}`);
  }

  const text = await response.text();
  if (!text) return undefined as T;
  return JSON.parse(text) as T;
}

interface PagedBitbucketResponse<T> {
  values: T[];
  size: number;
  isLastPage: boolean;
}

export interface ProjectResponse {
  id: number;
  key: string;
  name: string;
  description?: string;
  public?: boolean;
  type?: string;
}

export interface RepositoryResponse {
  id: number;
  slug: string;
  name: string;
  description?: string;
  project: { key: string; name: string };
  links: {
    clone?: Array<{ href: string; name: string }>;
    self?: Array<{ href: string }>;
  };
}

export interface BranchResponse {
  id: string;
  displayId: string;
  isDefault: boolean;
  latestCommit?: string;
}

export interface PullRequestResponse {
  id: number;
  title: string;
  state: string;
  links: { self: Array<{ href: string }> };
}

export async function listProjects(): Promise<ProjectResponse[]> {
  const response = await bitbucketFetch<PagedBitbucketResponse<ProjectResponse>>(
    "/projects?limit=100",
  );
  return response?.values ?? [];
}

export async function listRepositories(projectKey: string): Promise<RepositoryResponse[]> {
  const response = await bitbucketFetch<PagedBitbucketResponse<RepositoryResponse>>(
    `/projects/${encodeURIComponent(projectKey)}/repos?limit=100`,
  );
  /* v8 ignore start — defensive ?? when API returns null */
  return response?.values ?? [];
  /* v8 ignore stop */
}

export async function listAllRepositories(): Promise<RepositoryResponse[]> {
  const projects = await listProjects();
  const allRepos: RepositoryResponse[] = [];
  for (const project of projects) {
    const repos = await listRepositories(project.key);
    allRepos.push(...repos);
  }
  return allRepos;
}

export async function getRepository(
  projectKey: string,
  repoSlug: string,
): Promise<RepositoryResponse> {
  return bitbucketFetch<RepositoryResponse>(
    `/projects/${encodeURIComponent(projectKey)}/repos/${encodeURIComponent(repoSlug)}`,
  );
}

export async function getDefaultBranch(
  projectKey: string,
  repoSlug: string,
): Promise<BranchResponse> {
  return bitbucketFetch<BranchResponse>(
    `/projects/${encodeURIComponent(projectKey)}/repos/${encodeURIComponent(repoSlug)}/default-branch`,
  );
}

export async function listBranches(
  projectKey: string,
  repoSlug: string,
  filterText?: string,
): Promise<BranchResponse[]> {
  const filter = filterText ? `&filterText=${encodeURIComponent(filterText)}` : "";
  const response = await bitbucketFetch<PagedBitbucketResponse<BranchResponse>>(
    `/projects/${encodeURIComponent(projectKey)}/repos/${encodeURIComponent(repoSlug)}/branches?limit=100${filter}`,
  );
  return response?.values ?? [];
}

export async function createBranch(
  projectKey: string,
  repoSlug: string,
  branchName: string,
  startPoint: string,
): Promise<BranchResponse> {
  return bitbucketFetch<BranchResponse>(
    `/projects/${encodeURIComponent(projectKey)}/repos/${encodeURIComponent(repoSlug)}/branches`,
    {
      method: "POST",
      body: JSON.stringify({ name: branchName, startPoint }),
    },
  );
}

export async function createPullRequest(
  projectKey: string,
  repoSlug: string,
  title: string,
  description: string,
  sourceBranch: string,
  targetBranch: string,
  reviewers: string[] = [],
): Promise<PullRequestResponse> {
  const body = {
    title,
    description,
    fromRef: { id: `refs/heads/${sourceBranch}` },
    toRef: { id: `refs/heads/${targetBranch}` },
    reviewers: reviewers.map((r) => ({ user: { name: r } })),
  };

  return bitbucketFetch<PullRequestResponse>(
    `/projects/${encodeURIComponent(projectKey)}/repos/${encodeURIComponent(repoSlug)}/pull-requests`,
    { method: "POST", body: JSON.stringify(body) },
  );
}

export function getCloneUrl(projectKey: string, repoSlug: string): string {
  const { baseUrl } = getBitbucketConfig();
  return `${baseUrl}/scm/${projectKey.toLowerCase()}/${repoSlug}.git`;
}

export function getPullRequestWebUrl(
  projectKey: string,
  repoSlug: string,
  prId: number,
): string {
  const { baseUrl } = getBitbucketConfig();
  return `${baseUrl}/projects/${projectKey}/repos/${repoSlug}/pull-requests/${prId}`;
}
